File: packages/mlforge_sdk/mlforge_sdk/exports.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict

from .http import HttpClient


class ExportJob(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    project_id: str
    run_id: str
    format: str
    status: str
    progress: Optional[float] = None
    artifact_url: Optional[str] = None
    created_at: Optional[str] = None
    completed_at: Optional[str] = None


class ExportsClient:
    def __init__(self, http: HttpClient):
        self._http = http

    def list(
        self,
        *,
        project_id: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[ExportJob]:
        params: Dict[str, Any] = {"project_id": project_id, "limit": limit, "offset": offset}
        data = self._http.get("/exports", params=params)
        return [ExportJob(**e) for e in data]

    def get(self, export_id: str) -> ExportJob:
        data = self._http.get(f"/exports/{export_id}")
        return ExportJob(**data)

    def start(self, ctx: Dict[str, Any]) -> Dict[str, Any]:
        data = self._http.post("/exports", json_body=ctx)
        return data

    def cancel(self, export_id: str) -> None:
        self._http.post(f"/exports/{export_id}/cancel")

    def download(self, export_id: str) -> bytes:
        """Download export artifact as bytes."""
        import requests

        url = f"{self._http._base_url}/exports/{export_id}/download"
        token = getattr(self._http, "_token", None)
        headers = {"Authorization": f"Bearer {token}"} if token else {}

        resp = requests.get(url, headers=headers)
        resp.raise_for_status()
        return resp.content

File: packages/mlforge_js_sdk/src/exports.ts
import type { HttpClient } from "./http.js";

export interface ExportJob {
  id: string;
  project_id: string;
  run_id: string;
  format: string;
  status: string;
  progress?: number;
  artifact_url?: string;
  created_at?: string;
  completed_at?: string;
}

export interface ExportOptions {
  project_id: string;
  run_id: string;
  format: string;
  include_weights?: boolean;
  include_metadata?: boolean;
  [key: string]: unknown;
}

export class ExportsClient {
  constructor(private http: HttpClient) {}

  async list(opts?: {
    project_id?: string;
    limit?: number;
    offset?: number;
  }): Promise<ExportJob[]> {
    return this.http.get<ExportJob[]>("/exports", {
      project_id: opts?.project_id,
      limit: opts?.limit ?? 50,
      offset: opts?.offset ?? 0,
    } as Record<string, string | number | boolean>);
  }

  async get(exportId: string): Promise<ExportJob> {
    return this.http.get<ExportJob>(`/exports/${exportId}`);
  }

  async start(opts: ExportOptions): Promise<{ job_id: string; status: string }> {
    return this.http.post<{ job_id: string; status: string }>("/exports", opts);
  }

  async cancel(exportId: string): Promise<void> {
    return this.http.post<void>(`/exports/${exportId}/cancel`);
  }

  async download(exportId: string): Promise<Blob> {
    const response = await fetch(this.http["baseUrl"] + `/exports/${exportId}/download`);
    if (!response.ok) {
      throw new Error(`Export download failed: ${response.statusText}`);
    }
    return response.blob();
  }
}
